pkg/queue/breaker.go

```diff
@@ -177,131 +176,137 @@ func (b *Breaker) Capacity() int {
 	return b.sem.Capacity()
 }
 
-// newSemaphore creates a semaphore with the desired maximal and initial capacity.
-// Maximal capacity is the size of the buffered channel, it defines maximum number of tokens
-// in the rotation. Attempting to add more capacity then the max will result in error.
-// Initial capacity is the initial number of free tokens.
+// newSemaphore creates a semaphore with the desired initial capacity.
 func newSemaphore(maxCapacity, initialCapacity int) *semaphore {
 	queue := make(chan struct{}, maxCapacity)
 	sem := &semaphore{queue: queue}
-	if initialCapacity > 0 {
-		sem.updateCapacity(initialCapacity)
-	}
+	sem.updateCapacity(initialCapacity)
 	return sem
 }
 
-// semaphore is an implementation of a semaphore based on Go channels.
-// The presence of elements in the `queue` buffered channel correspond to available tokens.
-// Hence the max number of tokens to hand out equals to the size of the channel.
-// `capacity` defines the current number of tokens in the rotation.
+// semaphore is an implementation of a semaphore based on packed integers and a channel.
+// state is an uint64 that has two uint32s packed into it: capacity and inFlight. The
+// former specifies how many request are allowed at any given time into the semaphore
+// while the latter refers to the currently in-flight requests.
+// Packing them both into one uint64 allows us to optimize access semantics using atomic
+// operations, which can't be guaranteed on 2 individual values.
+// The channel is merely used as a vehicle to be able to "wake up" individual goroutines
```

Comment:
Thinking out loud, this description sort of makes me wonder if we should just use `sync.Condition` here, though maybe the channel is just as good a way of doing this

Reply:
The crucial difference is that `sync.Cond`'s wakeups need a corresponding waiting goroutine at that very moment. To guarantee that both waiters and releasers see the same state of a sync.Cond, you need to use it's locker which forces you to use a mutex basically.

The channel allows us to "delay" the wakeup in that one goroutine can send a wakeup to the channel and somewhen in the future, another goroutine can be woken up without the two having to synchronize on their state.